### PR TITLE
 Link nicks mentioned in messages 

### PR DESCRIPTION
**Pull request by MaxLeiter**
Not sure how I like this. kiwi does it. Can disable in settings (but requires reload to change sent messages, which is fine imo). Setting probably needs better wording.

Should be good to go.

screen shot 2017-11-14 at 6 50 54 pm

relies on #1712

### CODE REVIEW

**Comment by christer88** [comment]
Big thumbs up for this!

**Comment by MaxLeiter** [comment]
I think this is good to go

File: client/js/options.js [hunk 1]
```diff
@@ -18,14 +18,15 @@ const options = {
...
+	linkNicks: true,
 	motd: true,
 	notification: true,
 	notifyAllMessages: false,
 	showSeconds: false,
 	statusMessages: "condensed",
 	theme: $("#theme").attr("href").replace(/^themes\/(.*).css$/, "$1"), // Extracts default theme name, set on the server configuration
 	thumbnails: true,
-	userStyles: userStyles.text(),
+	userStyles: userStyles.text()
```
Comment (astorije) [inline client/js/options.js]: Revert this change

File: client/js/libs/handlebars/ircmessageparser/findNames.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"use strict";
+
+function findNames(text) {
```
Comment (astorije) [inline client/js/libs/handlebars/ircmessageparser/findNames.js]: Assuming this PR gets accepted overall, these 2 files + changes to `parse` could totally use some test cases :)
Don't worry about it for now though, let's get the rest of the PR settled first.

File: client/index.html [hunk 3]
```diff
@@ -221,6 +221,12 @@ <h1 class="title">Settings</h1>
 									Show seconds in timestamp
 								</label>
 							</div>
+							<div class="col-sm-7">
+								<label class="opt">
+									<input type="checkbox" name="linkNicks">
+									Make mentioned names clickable
+								</label>
+							</div>
```
Comment (astorije) [inline client/index.html]: I know this is gonna add up to my rep, but I'm not at all in favor of a user setting for this.
If this PR works and we want it (I personally do!), let's make some UX decision and simply make this a thing.
Code will be simpler, less branching (i.e. less things to make sure we do not break), and it's a good addition overall. 👍

**Comment by astorije** [comment]
This is a very fine addition, I haven't tested but love it already 👏 👏 Awesome job @MaxLeiter!

Last time this was attempted (client-side only), @xPaw has some performance concerns. The way you did it seems to me like you do not have this issue anymore. I'll let @xPaw give a first pass at this for this reason.

**Comment by MaxLeiter** [comment]
I'm all for no setting 👍

File: client/js/options.js (hunk 1 above)
Comment (MaxLeiter) [reply client/js/options.js]: Fixed

File: client/js/libs/handlebars/ircmessageparser/findNames.js (hunk 2 above)
Comment (MaxLeiter) [reply client/js/libs/handlebars/ircmessageparser/findNames.js]: 👍

File: client/js/libs/handlebars/ircmessageparser/parseName.js [hunk 4]
```diff
@@ -22,31 +20,26 @@ module.exports = function(orig) {
...
-				suffix = "'s";
-				strip++;
-			}
+	const prefixes = ["@", "+", "%", "!"];
+	prefixes.map((e) => {
+		if (orig.startsWith(e)) {
+			cleaned = cleaned.substring(e.length, cleaned.length);
+			prefix = e;
 		}
+	});
```
Comment (RJacksonm1) [inline client/js/libs/handlebars/ircmessageparser/parseName.js]: This prefix and suffix logic could be simplified in another way, without multiple `startsWith` checks:

```
const prefixes = ["@", "+", "%", "!"];

let prefix = "";
let matchedPrefixPos = prefixes.indexOf(orig[0]);

if (matchedPrefixPos !== -1) {
    prefix = prefixes[matchedPrefixPos];
    cleaned = cleaned.substring(0, cleaned.length - prefix.length);
}
```

File: src/plugins/irc-events/message.js [hunk 5]
```diff
@@ -87,14 +87,25 @@ module.exports = function(irc, network) {
 			highlight = network.highlightRegex.test(data.message);
 		}
 
+		const users = [];
+		const splitMessage = data.message.split(" ");
+		splitMessage.forEach((val, index) => {
+			const cleanedNick = val.replace(":", "").replace(",", "").replace("@", "").replace("'s", ""); // common nick prefixes/suffixes when chatting
+			const user = chan.findUser(cleanedNick);
```
Comment (xPaw) [inline src/plugins/irc-events/message.js]: This doesn't sound like a great idea to do it the channel has thousands of users and its highly active.

Comment (MaxLeiter) [reply src/plugins/irc-events/message.js]: Does the server have a concept of rendered messages on the client? I.e. when you click “show more”/on init? Could do it there

File: client/js/libs/handlebars/ircmessageparser/parseName.js (hunk 4 above)
Comment (MaxLeiter) [reply client/js/libs/handlebars/ircmessageparser/parseName.js]: won't work if the suffix/prefix we're looking for is length > 1 (due to `indexOf(orig[0])`)

**Comment by astorije** [comment]
@MaxLeiter, could you rebase on master please?

**Comment by astorije** [comment]
@MaxLeiter, since #1712 was merged, I took the liberty to rebase that for you and fix conflicts (which were just dangling comma shit) so I can test it with most recent master.

File: src/models/msg.js [hunk 6]
```diff
@@ -13,6 +13,7 @@ class Msg {
 			text: "",
 			type: Msg.Type.MESSAGE,
 			self: false,
+			users: [], // {start, end, nick}
```
Comment (xPaw) [inline src/models/msg.js]: Hm, I don't think we need to set a default? `parse` should be able to deal with it just fine.

Also, it might be beneficial to rename this to `nicks` to be more specific.

Comment (astorije) [reply src/models/msg.js]: I'm getting the following, which might be related to your comment:

```
lounge/src/models/chan.js:124
	return this.users.get(nick.toLowerCase());
	                          ^

TypeError: Cannot read property 'toLowerCase' of undefined
    at Chan.findUser (/Users/jeremie/Perso/lounge/src/models/chan.js:124:28)
    at Chan.getUser (/Users/jeremie/Perso/lounge/src/models/chan.js:128:14)
    at IrcClient.<anonymous> (/Users/jeremie/Perso/lounge/src/plugins/irc-events/topic.js:18:15)
    at IrcClient.emit (/Users/jeremie/Perso/lounge/node_modules/eventemitter3/index.js:129:35)
    at /Users/jeremie/Perso/lounge/node_modules/irc-framework/src/client.js:185:20
    at next (/Users/jeremie/Perso/lounge/node_modules/middleware-handler/index.js:42:18)
    at MiddlewareHandler.handle (/Users/jeremie/Perso/lounge/node_modules/middleware-handler/index.js:64:3)
    at IrcCommandHandler.<anonymous> (/Users/jeremie/Perso/lounge/node_modules/irc-framework/src/client.js:179:34)
    at IrcCommandHandler.emit (/Users/jeremie/Perso/lounge/node_modules/eventemitter3/index.js:130:35)
    at IrcCommandHandler.emit (/Users/jeremie/Perso/lounge/node_modules/irc-framework/src/commands/handler.js:95:23)
```

Comment (xPaw) [reply src/models/msg.js]: Doesn't appear related to the PR.

Comment (astorije) [reply src/models/msg.js]: It's not happening on `master`, so I'm guessing something gets passed wrong somewhere up the stack.

**Comment by astorije** [comment]
@xPaw, did you see this? https://travis-ci.org/thelounge/lounge/jobs/304356874#L2268
I'm getting this when testing this PR locally as well.

File: client/js/libs/handlebars/parse.js [hunk 7]
```diff
@@ -83,6 +91,8 @@ module.exports = function parse(text) {
 			return `<span class="inline-channel" role="button" tabindex="0" data-chan="${escapedChannel}">${fragments}</span>`;
 		} else if (textPart.emoji) {
 			return `<span class="emoji">${fragments}</span>`;
+		} else if (textPart.nick) {
+			return templates.user_name({nick: textPart.nick}).trim();
```
Comment (xPaw) [inline client/js/libs/handlebars/parse.js]: Unfortunately to fix tests (at least right now), the easiest solution would be copying user_name template here.

Tests fail because .tpl files aren't being compiled when required from mocha.

Also probably should be using `fragments` here like the others above.

File: src/models/msg.js (hunk 6 above)
Comment (MaxLeiter) [reply src/models/msg.js]: yeah, this works fine without the default

File: client/js/libs/handlebars/parse.js (hunk 7 above)
Comment (MaxLeiter) [reply client/js/libs/handlebars/parse.js]: Done

**Comment by astorije** [comment]
@MaxLeiter, @xPaw, I believe I have something to fix the failing test. I need to play more with that though to make it merge-ready. I'll try to do that after work tonight. Thus ignore it for now.

File: client/js/libs/handlebars/parse.js [hunk 8]
```diff
@@ -83,6 +91,9 @@ module.exports = function parse(text) {
 			return `<span class="inline-channel" role="button" tabindex="0" data-chan="${escapedChannel}">${fragments}</span>`;
 		} else if (textPart.emoji) {
 			return `<span class="emoji">${fragments}</span>`;
+		} else if (textPart.nick) {
+			const nick = fragments.trim();
```
Comment (xPaw) [inline client/js/libs/handlebars/parse.js]: Why is it trimmed?

File: client/js/libs/handlebars/parse.js [hunk 9]
```diff
@@ -83,6 +91,9 @@ module.exports = function parse(text) {
...
+			return `<span role="button" class="user ${colorClass(nick)}" data-name="${nick}">${nick}</span>`;
```
Comment (xPaw) [inline client/js/libs/handlebars/parse.js]: `fragments` may contain HTML, so you probably want to use `textPart.nick` in `colorClass`/`date-name` (and escape it?)

File: client/js/libs/handlebars/parse.js (hunk 8 above)
Comment (MaxLeiter) [reply client/js/libs/handlebars/parse.js]: Was an issue from before with handlebars, need to test if necessary now

File: client/js/libs/handlebars/parse.js (hunk 9 above)
Comment (MaxLeiter) [reply client/js/libs/handlebars/parse.js]: Don't think I need to escape?

**Comment by MaxLeiter** [comment]
@Bonuspunkt could you take a look? This is broken with nicks, colors, bold, etc

**Comment by Bonuspunkt** [comment]
i'm guessing rel #1718 ?
what are the expected results?

**Comment by MaxLeiter** [comment]
Right now duplicates are produced (www.MaxLeiter.com with nick MaxLeiter in Chan renders www.MaxLeiter.comMaxLeiter. Each word is correct (link is a link, nick is nick))

**Comment by astorije** [comment]
FYI, just tested it works fine with copy/paste. Can't wait for this feature!

**Comment by xPaw** [comment]
@MaxLeiter Can you add some tests for this? Like `http://example.com` with ` example` being an user, and some others?

File: test/client/js/libs/handlebars/parse.js [hunk 10]
```diff
@@ -243,6 +243,49 @@ describe("parse Handlebars helper", () => {
 		expect(actual).to.deep.equal(expected);
 	});
 
+	it("should find nicks", () => {
+		const testCases = [{
+			users: ["MaxLeiter"],
+			input: "test, MaxLeiter",
+			expected:
+				"test, " +
+				"<span role=\"button\" class=\"user color-12\" data-name=\"MaxLeiter\">" +
```
Comment (astorije) [inline test/client/js/libs/handlebars/parse.js]: Can you use single quotes instead to avoid escaping? (See my PR I just merged)

Comment (xPaw) [reply test/client/js/libs/handlebars/parse.js]: This was merged before yours. You should have enforced this with eslint rule by the way.

**Comment by astorije** [comment]
Yeah sorry I was on phone and didn't realize it was merged.
I don't think there is a way to enforce it, just allow it, no?
-- 
Jérémie Astori
https://jeremie.astori.fr
